### ISSUE TITLE
stack build in WSL - openAnonymousTempFileFromDir: inappropriate type (Is a directory)

### ISSUE THREAD

**Issue by razvan-panda** [commercialhaskell/stack#5010]
### Steps to reproduce

* have Ubuntu 18.04 WSL
* install Nix in it
* git clone https://github.com/fairy-tale-agi-solutions/fairy-tale-artificial-general-intelligence-solutions.git
* checkout revision 6354fcca2bb37ad4f4c6b05fdcb55b09734f5164
* run `nix-shell` in project root
* run `stack build --verbose`

### Expected

Project to build successfully.

### Actual

```
stack build --verbose
2.1.3.1 x86_64 hpack-0.31.2
2019-08-19 14:39:27.610378: [debug] Checking for project config at: /home/neo/Projects/fairy-tale-artificial-general-intelligence-solutions/Sources/Razvan Flavius Panda/Haskell/artificial-general-intelligence/stack.yaml
2019-08-19 14:39:27.611197: [debug] Loading project config file stack.yaml
2019-08-19 14:39:27.612420: [error] openAnonymousTempFileFromDir: inappropriate type (Is a directory)
```

### Stack version

```
$ stack --version
2.1.3.1 x86_64 hpack-0.31.2
```

### Method of installation

Installed stack via Nix.

**Comment by razvan-panda** [comment]
I can't reproduce the error anymore. Builds successfully now.

**Comment by razvan-panda** [comment]
I got the same issue on another project when trying to debug using: https://github.com/phoityne/phoityne-vscode

Error log:

```
CWD: /home/neo/Projects/haskell-stack-fix
CMD: stack ghci --with-ghc=haskell-dap --test --no-load --no-build --main-is TARGET --ghci-options -fprint-evld-with-show
Writing implicit global project config file to: /home/neo/.stack/global-project/stack.yaml
Note: You can change the snapshot via the resolver field there.
Using latest snapshot resolver: lts-14.1
openAnonymousTempFileFromDir: inappropriate type (Is a directory)

------------------------------------
can not parse ghci version. ["getGHCiVersion" (line 5, column 1):
unexpected end of input
expecting "GHCi, version "] assumes 8.0.0.
------------------------------------
:set prompt "H>>= "
```

**Comment by dbaynard** [comment]
Well [`openAnonymousTempFileFromDir` comes from `unliftio`](https://github.com/fpco/unliftio/blob/40e0f72dc0f79dbecb103a1bc6c27fff36251735/unliftio/src/UnliftIO/IO/File/Posix.hs#L218). Looks like the issue could be that the filepath provided [lacks a trailing slash](https://github.com/haskell/filepath/blob/6c11507fa7449d8d2ad1d0286d225bbe169b5816/System/FilePath/Internal.hs#L515-L526). Or perhaps the directory is being passed where a file is expected.

It's odd that it's intermittent, though. I wonder whether it's a WSL bug.

**Comment by pepeiborra** [comment]
I get this issue when building inside a nix-shell on RHEL6, so not WSL specific.

**Comment by sgalal** [comment]
The same problem for me. I am using Arch Linux in WSL. I installed the latest version of stack via pacman.

**Comment by ju1m** [comment]
Same problem on Debian jessie using stack from nixpkgs-19.09 (stack 2.1.3.1 x86_64 hpack-0.32.0).
The problem fails further along the line (stack now downloads the index) when using deprecated paths: `~/.stack/stack.yaml` instead of `~/.stack/config.yaml` and `~/.stack/global` instead of `~/.stack/global-project`.
stack tries `openat()` on the directories `~/.stack` or `~/.stack/global`.

**Comment by buwailee** [comment]
A possible solution: try wsl2

**Comment by pikajude** [comment]
Unfortunately WSL2 isn't a viable solution to anything, not until they fix the unbounded memory leak problem :/

**Comment by ncihnegn** [comment]
Same issue on Arch WSL 1

**Comment by NickSeagull** [comment]
Just adding my +1

**Comment by ahri** [comment]
Just a note; the ubuntu-packaged version works, so this worked for me:

    $ sudo apt install haskell-stack && stack upgrade

**Comment by NickSeagull** [comment]
Yep, it works from the official installer, the issue is with the Nix one, so perhaps we should open this ticket in `nixpkgs`?

**Comment by sgalal** [comment]
> Yep, it works from the official installer, the issue is with the Nix one, so perhaps we should open this ticket in `nixpkgs`?

No. As I commented before, it is not a Nix-only issue.

**Comment by Neurognostic** [comment]
Seeing the same issue in Arch on WSL 1.

**Comment by Neurognostic** [comment]
Found a workaround.

**tldr:**
```shell
pacman -S stack
mkdir -p ~/.stack
echo {} > ~/.stack/config.yaml
/usr/sbin/stack upgrade --force-download
stack build
```

---

With the `stack` package downloaded from the Arch [Community repository](https://www.archlinux.org/packages/community/x86_64/stack/) my projects fail to build:

```console
$ pacman -S stack
$ /usr/sbin/stack --version
Version 2.1.3, Git revision b2e2f3d0568e6d73fc0477dee8ca7c04466ec1f6 (dirty) (7736 commits) x86_64
Compiled with:
- Cabal-2.4.0.1
- Glob-0.10.0
...
- zip-archive-0.4.1
- zlib-0.6.2.1

Warning: this is an unsupported build that may use different versions of
dependencies and GHC than the officially released binaries, and therefore may
not behave identically.  If you encounter problems, please try the latest
official build by running 'stack upgrade --force-download'.
```

When building with this version the error is shown. If I try to follow the above warning's guidance and to install an "official release" the error is shown again.

```console
[error] openAnonymousTempFileFromDir: inappropriate type (Is a directory)
```

Following @pikajude's suggestion to create `~/.stack/config.yml` with only `{}` still gave me the error when building my project, but it did allow me to install the "official" `stack` release.

```shell
mkdir -p ~/.stack
echo {} > ~/.stack/config.yaml
```

> **NOTE:** `config.yml` does **not** work. The same error is shown. Must use `.yaml` extension.

```shell
# Install official build
stack upgrade --force-download

# Check new version / git revision
/usr/sbin/stack --version
Version 2.1.3, Git revision 636e3a759d51127df2b62f90772def126cdf6d1f (7735 commits) x86_64 hpack-0.31.2
```

Now my projects **build** successfully.

CC: @felixonmars ( Arch [`stack` package](https://www.archlinux.org/packages/community/x86_64/stack/) maintainer)

**Comment by Neurognostic** [comment]
The AUR `stack-bin` package also works.

```shell
git clone https://aur.archlinux.org/packages/stack-bin
cd stack-bin
makepkg --syncdeps --clean
```

```console
$ /usr/sbin/stack --version
Version 2.1.3, Git revision 0fa51b9925decd937e4a993ad90cb686f88fa282 (7739 commits) x86_64 hpack-0.31.2
```

**Comment by hongy19** [comment]
I added a Archlinux bug on this issue https://bugs.archlinux.org/task/65263

**Comment by eli-schwartz** [comment]
Don't report bugs for your frankenstein Microsoft system. Arch Linux does not recognize WSL as valid Arch Linux and we don't provide support for it.

We are not your support channel for the fact that Microsoft's WSL emulation layer is incomplete and doesn't work 100%.

...

Especially since you claimed in your Arch Linux bug report that "but other distribution package and The AUR stack-bin package doesn't has this issue", when this github issue quite plainly demonstrates that other distribution packages *do* have the issue... on WSL... even if the upstream prebuilt binaries do not.

**Comment by carlosdagos** [comment]
Had the same issue using `nix` in `Ubuntu-20.04` in WSL1, I did part of what @Neurognostic has above.

```
mkdir -p ~/.stack
echo {} > ~/.stack/config.yaml
```

The error went away for the initial stages (downloading the index), but came back whilst trying to actually build a project and I couldn't find a way around it.

My solution for now: I downloaded the statically linked binary from the releases page. That one seems to be doing the work smoothly.

For anybody else making the dubious but courageous effort to use WSL1 mixed with `nix`, so far what seems to be the smoothest: 
* Use the statically linked `stack` binary
* Use the Nix integration as described here: https://docs.haskellstack.org/en/stable/nix_integration/#additions-to-your-stackyaml

By enabling nix in the stack config, you will just inherit a bunch of stuff required to configure packages (`gcc`, `make`, `gmp`), which don't come with Ubuntu 20.04 by default.

Having said that, it seems that WSL is full of fun times 😄

**Comment by and-pete** [comment]
Also received `openAnonymousTempFileFromDir: inappropriate type (Is a directory)` when installing `ghc-8.8.3` via `stack` on Ubuntu 18.04 on WSL1.

What's different for me compared to many comments above is that I was not using `nix` at all. Just straight Ubuntu.

`stack --version` gives `Version 2.4.0, Git revision 9dcef52902d01646d63fe76fc8e6b1b3ac6cc9b8` from using `stack upgrade --git`

**Comment by dten** [comment]
encountered this recently and tracked it to this change. reverting the changes to Prelude here make the error go away.
https://github.com/commercialhaskell/stack/commit/62c24fa0552dc722260f1cee75325145e837ab6a#diff-748b3631a8e26a9b793853e343ca0577

**Comment by dten** [comment]
What i don't understand is why the binary from `stack upgrade --binary-version 2.3.1` works fine but building the tag v2.3.1 creates a stack can't build anything because of the error in this issue
i'm on Windows 2004 with WSL 1 runng Ubuntu 18.04 by the way

**Comment by aryairani** [comment]
I tried this:
```
mkdir -p ~/.stack
echo {} > ~/.stack/config.yaml
```
and then my build started, looking normal, but causes my system to reboot during the build. Sorry I'm not leaving much system config here due to laziness, but I wanted to mention the reboot / possible BSOD.
